src/helpers/data_channel.ts

```diff
@@ -0,0 +1,146 @@
+        this._ready.resolve();
+
+        // Unpause once low water mark has been reached
+        this.dc.bufferedAmountLowThreshold = lowWaterMark;
+        this.dc.onbufferedamountlow = () => {
+            if (!this._ready.done) {
+                this.$log.debug(this.logTag, `${this.dc.label} resumed (buffered=${this.dc.bufferedAmount})`);
+                this._ready.resolve();
+            }
+        };
+    }
+
+    /**
+     * A future whether the data channel is ready to be written on.
+     */
+    public get ready(): Future<void> {
+        return this._ready;
+    }
+
+    /**
+     * Write a message to the data channel's internal buffer for delivery to
+     * the remote side.
+     *
+     * Important: Before calling this, the `ready` Promise must be awaited.
+     *
+     * @param message The message to be sent.
+     * @throws Error in case the data channel is currently paused.
+     */
+    public write(message: Uint8Array): void {
+        // Throw if paused
+        if (!this._ready.done) {
+            throw new Error('Unable to write, data channel is paused!');
+        }
+
+        // Try sending
+        // Note: Technically we should be able to catch a TypeError in case the
+        //       underlying buffer is full. However, there are other reasons
+        //       that can result in a TypeError and no browser has implemented
+        //       this properly so far. Thus, we use a well-tested high water
+        //       mark instead and try to never fill the buffer completely.
+        this.dc.send(message);
+
+        // Pause once high water mark has been reached
+        if (this.dc.bufferedAmount >= this.highWaterMark) {
+            this._ready = new Future();
+            this.$log.debug(this.logTag, `${this.dc.label} paused (buffered=${this.dc.bufferedAmount})`);
+        }
+    }
+}
+
+/**
+ * A flow-controlled (sender side) data channel that allows to queue an
+ * infinite amount of messages.
+ *
+ * While this cancels the effect of the flow control, it prevents the data
+ * channel's underlying buffer from becoming saturated by queueing all messages
+ * in application space.
+ */
+export class UnboundedFlowControlledDataChannel extends FlowControlledDataChannel {
+    private queue: Promise<void>;
+
+    /**
+     * Create a flow-controlled (sender side) data channel with an infinite
+     * buffer.
+     *
+     * @param $log The Angular logging service
+     * @param dc The data channel to be flow-controlled
+     * @param lowWaterMark The low water mark unpauses the data channel once
+     *   the buffered amount of bytes becomes less or equal to it.
+     * @param highWaterMark The high water mark pauses the data channel once
+     *   the buffered amount of bytes becomes greater or equal to it.
+     */
+    public constructor($log: ng.ILogService, dc: RTCDataChannel, lowWaterMark?: number, highWaterMark?: number) {
+        super($log, dc, lowWaterMark, highWaterMark);
+        this.queue = this.ready;
```

Comment:
Again, when `this._ready` is replaced with a new future, won't the queue still reference the old future? Could this be a problem?

Probably I'm missing something, but as far as I can see, the future is never reset, but replaced. So once the promise in the root of the queue is resolved it would stay in resolved state forever.

Reply:
Yes, in this case it references the `Future` that has been created on construction (see the constructor of `FlowControlledDataChannel`). That is the starting point of the queue. With each call to `write`, the `queue` is awaited by the `.writeWhenReady` `Promise` and immediately replaced by that promise.

Basically, `queue` is always the `Future` of the most recent `.write` call and resolves once it has written.